### ISSUE TITLE
Notes for noise scheduler in diffusion model

### ISSUE THREAD

**Issue by pomelyu** [pomelyu/paper-reading-notes#24]
#### 1. propose the definition of forward equation. i.e.
<img width="1010" alt="image" src="https://github.com/user-attachments/assets/5bb2083e-28c3-4963-a922-68861bcb5fb7">
and thus <img width="1034" alt="image" src="https://github.com/user-attachments/assets/d32c2225-0697-40b4-adbd-d062f7bc3d22">

#### 2. design a neutral network that has the below property to approximate the reverse(denoising) process
<img width="1008" alt="image" src="https://github.com/user-attachments/assets/c4c46fd6-81f2-4e20-b87d-f42e9af79164">
then we have
<img width="1015" alt="image" src="https://github.com/user-attachments/assets/9848091a-279a-497e-b2bb-90266f99f1e9">

#### 3. If we further assume $\Sigma_\theta(x_t, t) = \sigma_t^2$, we can get
<img width="1021" alt="image" src="https://github.com/user-attachments/assets/c60e5892-3d0d-4fa0-bec9-41b0a40280bd">
The formula indicates that we can design the network to predict "noise" instead of the "denoised sample". Finally we get the DDPM algorithm
<img width="1008" alt="image" src="https://github.com/user-attachments/assets/de22c8a8-d642-48e5-a16d-606c3655cca4">

Or, we can use formula in 2. to estimate $x_0$  first and then derive $x_{t-1}$ by $q(x_{t-1}|x_0, x_t)$
<img width="1016" alt="image" src="https://github.com/user-attachments/assets/ccd901f4-403c-4163-b2ce-1b880988bd04">

#### 7. In DDPM, there is a scale function(depends on t) to scale the loss value in training for every time step

**Comment by pomelyu** [comment]
## DDIM
#### 1. Design a non-Markovian inference models that lead to the same training objective in DDPM, but the reverse process is now deterministic. 
<img width="1032" alt="image" src="https://github.com/user-attachments/assets/e40384f2-2df7-4610-9bf4-7cbcbf482867">
and thus
<img width="1033" alt="image" src="https://github.com/user-attachments/assets/7c38255d-61f8-4202-9102-b5755f7aafd5">

When $\sigma_t=1$, DDIM becomes DDPM
#### 2. The denoising process can be reduced to about 50 steps

**Comment by pomelyu** [comment]
## Scored-based Generative Modeling through Stochastic Differential Equation
#### 1. prove that we can use a stochastic differential equation(SDE) to describe the process of diffusion probability model(DPM). i.e the solution of the SDE in each time step t is the sample in DPM in that time. Note the time in SDE is continuous
Forward process
<img width="1137" alt="image" src="https://github.com/user-attachments/assets/c040d06d-d7df-41b7-abdf-81d03846e264">
Then we have reverse process
<img width="1195" alt="image" src="https://github.com/user-attachments/assets/24ee5400-29b3-49fa-bbc8-e1e9b3f8717e">

$\mathcal{w}_t$ and $\bar{\mathcal{w}}_t$ are the standard Wiener process(Brownian motion) and reverse Wiener process respectively. We can get denoising samples at time step t by solving the reverse SDE at that time. $\nabla_x \log q_t(\mathbf{x}_t)$ is *score function* and it's the only unknown term in the reverse SDE. As a result, we can use a neural network to formulate it.
#### 2. Further, the SDE can be turned to Ordinary Differential Equation(ODE) if we formulate the sample distribution instead of a sample, SDE in reverse process becomes
<img width="1186" alt="image" src="https://github.com/user-attachments/assets/531ec154-1508-4fc2-978b-8ca3059ede72">

As mentioned previously, $\nabla_x \log q_t(\mathbf{x}_t)$ can be formulated by a neural network
#### 3. As a result, we can use general purpose method to solve this ODE, such as Euler method
#### 4. Indeed, this paper propose a unified frame work for DPM and SMLD(Score matching with Langevin Dynamics)

**Comment by pomelyu** [comment]
## (EDM)Elucidating the Design Space of Diffusion-Based Generative Models
#### 1. training, denoising process(sampling), model architecture can be decoupled
#### 2. propose modified step size selection, which makes step size smaller when t is approaching 0
<img width="1035" alt="image" src="https://github.com/user-attachments/assets/e01e585f-762b-4b4a-a5c6-b0e65d14c8f5">

#### 3. high order solver(e.g. Heun's) for ODE is better than first order(Euler), since the local error in every time step will accumulate.

**Comment by pomelyu** [comment]
## DPM-Solver
#### 1. propose dedicated ODE solver for diffusion process, the results are deterministic.
From *Scored-based Generative Modeling through Stochastic Differential Equation*, we have
<img width="1177" alt="image" src="https://github.com/user-attachments/assets/2afebe75-424c-43f0-a122-5b551d621e5f">
Then we can get its solutions
<img width="1195" alt="image" src="https://github.com/user-attachments/assets/97adaf2a-c30f-4798-bb35-1f8ec57ce345">
<img width="1198" alt="image" src="https://github.com/user-attachments/assets/cc3bfaee-d77a-4c2a-8b81-67537112643a">
different "order" means using the different order in Taylor expansion to approximate integral
#### 2. DPM-solver reduce the sampling steps to around 10 steps(20 function calls when using 2nd solver)

**Comment by pomelyu** [comment]
## DPM-Solver++
#### 1. Find that the large guidance scale causes high-order solvers(includes DPM-Solver) to be instable in guided sampling, especially the one without latent diffusion
The author says that the larger guidance scale will amplify the derivatives of the model and thus the converge range of ODE solvers.
#### 2. The larger guidance scale also push the predicted noise away from the true noise and thus causes saturated and unnatural results.
#### 3. The author found this problem doesn't appear in stable diffusion, perhaps due to its powerful latent decoder 
#### 4. like DPM-solver, but the model predict $x_\theta$ instead of $\epsilon_\theta$, since we can adopt dynamic thresholding methods to mitigate the train-test mismatch problem
dynamic thresholding methods: Photorealistic text-to-image diffusion models with deep language understanding
#### 5. propose both single step and multi-step solver, they found 2nd-multi-step solver perform best
<img width="1009" alt="image" src="https://github.com/user-attachments/assets/072997a2-a476-4132-93bc-a8c63e5e6fa7">


---
DPM-Solver++ has a great interpretation about DPM, SDE and ODE for both $x_\theta$ and $\epsilon_\theta$ model

<img width="1011" alt="image" src="https://github.com/user-attachments/assets/884760aa-5bef-463a-a627-84796187495a">

**Comment by pomelyu** [comment]
## [PLMS,PNMS] Pseudo Numerical Methods for Diffusion Models on Manifolds
#### 1. Another sample scheduler based on high order ODE solver, which can reduce time steps to about 50
#### 2. According to DDIM, we have 
<img width="1002" alt="image" src="https://github.com/user-attachments/assets/6c9838a2-42d9-4abd-8992-20e27be040fa">
then
<img width="1031" alt="image" src="https://github.com/user-attachments/assets/07c72a81-4b5d-4172-9a2c-59dd3c7d9202">
<img width="990" alt="image" src="https://github.com/user-attachments/assets/0e3c4410-6620-467e-a4b7-d23aa2fb743a">

and $x_{t-\delta}$
<img width="1005" alt="image" src="https://github.com/user-attachments/assets/2cf70f15-40e9-4163-8f00-f4f63251c620">
If we replace $\epsilon$ by $\epsilon_\theta$, we have DDIM.
#### 3. We can use standard numerical method to get more precise $\epsilon$ from $\epsilon_\theta$, such as
*Linear Multi-Step Method*
<img width="827" alt="image" src="https://github.com/user-attachments/assets/e2e701bb-0974-42ec-9b2b-8d399a7de21a">
The author calls using high order numerical method for precise $\epsilon$ as *PNMS*
#### 4. The paper prove treating DDPMs as ODEs directly is not proper and has theoretical weakness, since the linear beta scheduler tend to let the equation has infinite derivatives.

**Comment by pomelyu** [comment]
## UniPC: A Unified Predictor-Corrector Framework for Fast Sampling of Diffusion Models
#### 1. UniC- $\rho$ can be added to any noise scheduler to correct the errors to add $\rho+1$ order accuracy (~high order ODE solver)
Theoretically, we can derive $x_t$ in reverse process by solving the following equation
<img width="1183" alt="image" src="https://github.com/user-attachments/assets/469ae3e6-d92c-470f-89ee-709c331109c4">
From DPM-Solver, we have approximate the solution by taylor expansion and have DDPM-Solver-1(~DDIM)
<img width="1189" alt="image" src="https://github.com/user-attachments/assets/8ebf3fa9-0cf0-492d-820e-8fa213806f21">
We can add an error correction term called UniC
<img width="1195" alt="image" src="https://github.com/user-attachments/assets/53efeb00-547a-4888-8ce3-e00cf353af7f">
Compare it with expanding the exponential integrator in (2). we have 
<img width="938" alt="image" src="https://github.com/user-attachments/assets/e52714a9-08c8-475b-80fa-3b1a4ee686a8">
<img width="1174" alt="image" src="https://github.com/user-attachments/assets/a781bc66-9b22-442b-8630-b2fd4d2fe219">

#### 2. UniP- $\rho$ is derived from UniC without estimating $\tilde{x}_{t_i}$ first.
It directly predicts $\tilde{x}_{t_i}$ from previous estimation
<img width="899" alt="image" src="https://github.com/user-attachments/assets/4d51659f-1561-40fb-8e5f-4b2fea96e3db">

#### 3. UniPC is the combination of UniC and UniP. It reduces reverse time steps to ~10 NFE(numbers of function evaluations)
<img width="1017" alt="image" src="https://github.com/user-attachments/assets/006ec346-e0dd-4f37-8ac4-628189ad1f94">
